### DISCUSSION TITLE
Using Picolibc with FreeRTOS 11.0.1

### DISCUSSION THREAD

**Discussion by tipeter** [picolibc/picolibc#688] [Q&A] [answered]
Greetings!

I want to use FreeRTOS 11 (11.0.1) kernel with stm32f439zi microcontroller in a simple CubeIDE project.
I don't use CMSIS-RTOS wrapper, only the native FreeRTOS API. I'm using the LL version of the STM32Cube drivers, as always.
The project has two versions:
- the version compiled with STM32's newlib-nano c library, linked with STM32's linker script
- and the version using the Picolibc library, using the Picolibc linker script

The newlib-nano version runs perfectly.
The Picolibc version crashes with HardFault in the "LL_RCC_HSE_IsReady()" function when running "SystemCoreClock_config()".

The relevant options in FreeRTOSConfig.h:
```
#define PICOLIBC_TLS
#define configUSE_PICOLIBC_TLS 1
```
Configuration of picolibc.ld:
```
__flash = 0x08000000;
__flash_size = 2M;
__ram = 0x20000000;
__ram_size = 192k;
__ccmram = 0x10000000;
__ccmram_size = 64k;
__stack_size = 1k;
```

To make it work for FreeRTOS, I recompiled Picolibc 1.8.6 with the "-Dthread-local-storage=true" option.
The toolchain used are gnu tools for stm32 12.3.rel1 and arm gnu toolchain 13.2.Rel1. 
Of course I used Picolibc compiled with the given toolchain. For both toolchains the same result.
The project builds without error, but the MCU stops with the HardFault error mentioned above.

**My specific question would be: how to correctly configure FreeRTOS to work with Picolibc?**

I've done a few smaller or bigger FreeRTOS projects, but in all of them I used the "factory" newlib-nano. This is the first time I've tried Picolibc.
To be honest the whole TLS issue is a bit unclear to me.

Any help is much appreciated in advance!

**Comment by keith-packard** [comment, marked as answer]
One issue is that the arm gnu toolchain still disables thread local storage support for some reason. I haven't tried FreeRTOS with that toolchain because of this. I maintain the Debian toolchain, which is exactly the same compiler source but built with TLS enabled, if you can use that.

Alternatively, you can try a toolchain built using crosstool-ng, which also leaves thread local storage enabled by default.

Thread local storage allows each thread to have its own version of "global" variables. The C library needs a couple of these due to its API design. The most obvious is 'errno'. Thread local storage is a standard mechanism where these variables are collected into a block of memory and referenced by a per-thread pointer.  The linker assigns offsets within this block and the compiler takes that offset and adds it to a per-thread pointer to access the variable. On 32-bit ARM targets, that per-thread pointer is fetched using an EABI-defined function, `__aeabi_read_tp`.

 This is a more efficient mechanism than used by the legacy newlib C library; that uses a giant struct for each thread which includes every global value used in the entire C library, not just the few values your application is likely to use.

The runtime needs to allocate space for this block in each thread and then ensure that the `__aeabi_read_tp` function will return the base address of the block for the current thread. The picolibc hooks in FreeRTOS allocate space for these variables at the top of each thread's stack area, and then use the picolibc support routines `_init_tls` and `_set_tls` to manage them.

None of this is likely to help you figure out what is actually broken though. I'd encourage you to try a compiler that doesn't disable TLS support and see if that helps though. The relevant GCC configuration option, `--disable-tls`, has no effect on code which doesn't use thread local storage; I don't understand why the ARM toolchain continues to use it, although I've filed a bug about it.

**Comment by tipeter** [comment]
Thank you for the explanation!

In the meantime, it turned out that the problem was caused by the STM32 "SystemInit()" function not running when using Picolibc. This function is called by Reset_Handler at the beginning of startup, but with Picolibc it runs its own crt0, not the STM32's startup_stm32f439zitx.s.
This wouldn't be a problem yet, but in my small test projects I don't load programs into Flash but into RAM. I use an appropriately modified picolibc.ld for this purpose. So I also move the interrupt vector table to the front of RAM, but the SCB->VTOR is set in SystemInit. This was not done for Picolibc.
As a temporary solution, I set VTOR at the beginning of "main()" and now everything works!

**Reply by keith-packard** [reply]
Glad you got it working! You should also be able to just use the STM32 startup code if you want; just need to copy a few bits from the picolibc code over to get things working.

**Reply by tipeter** [reply]
One more question: can CMake be used for the build? For me it would be more convenient than meson.
There is no mention of it in the documentation, but the changelog refers to it.
To tell you the truth I have successfully built with CMake, but I would feel safer with the "official" instructions :)
Sorry for the very trivial questions, I'm not an expert, just an amateur...

**Reply by keith-packard** [reply]
The CMake support exists solely to allow picolibc to be built as a Zephyr module. That tests much of the core library but not any of the support code (crt0, TLS, link scripts, etc). I'd recommend against using it; I realize yet-another-build-system is somewhat annoying.

**Reply by tipeter** [reply]
No problem, totally fine on meson too :)
Thanks for the help, and of course thanks for all the work you put into the Picolibc! :)
